### PR TITLE
Update Builder widget documentation

### LINKED ISSUE #48137: The use of term "platonic" is confusing
Internal: b/147018208

A number of docs for Flutter widgets in the API docs use the term "platonic" without defining the term, for example the Builder documentation: "A platonic widget that calls a closure to obtain its child widget."

It seems like this term is very confusing to external users, eg. https://stackoverflow.com/questions/52088889/can-someone-explain-to-me-what-the-builder-class-does-in-flutter
The term is also very confusing to me and my teammates, it took us quite some time to figure out what is meant by it.
It doesn't have any obvious connection to any dictionary definition of "platonic" that I could find, eg. https://www.merriam-webster.com/dictionary/platonic
It doesn't seem to be a term generally used in the CS or Software Engineering communities.
The only public mention I could find was in the Gitter link in one of the StackOverflow comments, https://gitter.im/flutter/flutter/archives/2017/05/02
I suspect that we invented the term internally and accidentally started using it in public documentation, overlooking that has entirely different meanings to an external audience.

Some possible solutions:

- Update the docs to explain what is meant in plain English, avoiding jargon, especially undefined jargon. For example, AnimatedBuilder, which I think is internally considered a "platonic" widget, is described as "A general-purpose widget for building animations.". Maybe the term "general-purpose" is a good replacement for platonic everywhere where platonic appears in the docs. If that sounds good, I'd be happy to prepare a PR into the public GitHub project.
- Invent some new Flutter jargon to describe the concept and define it clearly, since platonic already means something else for most English speakers.
- Remove the term "platonic" completely, eg. "A widget that calls a closure to obtain its child widget."

Thanks for taking a look at this!

List of affected docs:

https://api.flutter.dev/flutter/widgets/StatefulBuilder-class.html
https://api.flutter.dev/flutter/widgets/Builder-class.html

### PR DESCRIPTION
**Pull request by GroovinChip**
Also add myself to authors file.

This PR addresses @timsneath's request [on Twitter](https://twitter.com/timsneath/status/1291153383362043904?s=21) for documentation updates for the Builder widget to make it more readable and understandable. Addresses #48137. See [This StackOverflow post](https://stackoverflow.com/questions/52088889/can-someone-explain-to-me-what-the-builder-class-does-in-flutter) for more detail about the community's confusion as well.

## Related Issues

Fixes #48137

## Tests

No tests per se, as this PR is uniquely documentation based. I did run dartdoc to generate the documentation so I can see how it looks, though.

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I signed the [CLA].
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] All existing and new tests are passing.
- [ ] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [X] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [X] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by GroovinChip** [comment]
Running `flutter analyze --flutter-repo` produces 2 issues, not sure why. Here is the output:

```
Analyzing 3 directories...                                              

  error • A value of type 'Future<Dwds> Function({AssetReader assetReader, Stream<BuildResult> buildResults, Future<ChromeConnection> Function() chromeConnection, bool enableDebugExtension, bool enableDebugging, ExpressionCompiler expressionCompiler, String hostname, LoadStrategy loadStrategy, void
         Function(Level, String) logWriter, bool restoreBreakpoints, bool serveDevTools, Future<String> Function(String) urlEncoder, bool useSseForDebugProxy, bool verbose})' can't be assigned to a variable of type 'Future<Dwds> Function({AssetReader assetReader, Stream<BuildResult> buildResults,
         Future<ChromeConnection> Function() chromeConnection, bool enableDebugExtension, bool enableDebugging, ExpressionCompiler expressionCompiler, String hostname, LoadStrategy loadStrategy, void Function(Level, String) logWriter, bool serveDevTools, Future<String> Function(String) urlEncoder, bool
         useFileProvider, bool useSseForDebugBackend, bool useSseForDebugProxy, bool verbose})' • ../../../../../flutter/packages/flutter_tools/lib/src/build_runner/devfs_web.dart:152:33 • invalid_assignment
   info • The method doesn't override an inherited method • ../../../../../flutter/packages/flutter_tools/lib/src/build_runner/devfs_web.dart:603:18 • override_on_non_overriding_member

2 issues found. (ran in 70.8s)
```

File: packages/flutter/lib/src/widgets/basic.dart [hunk 1]
```diff
@@ -7160,13 +7160,88 @@ class KeyedSubtree extends StatelessWidget {
...
+///    );
+///  }
+/// ```
+/// {@end-tool}
+///
+/// At first glance this code might look fine, but would in fact not actually
+/// show the [SnackBar]. Since the [RaisedButton] is created in the same build
+/// method as the [Scaffold], the context object called in `Scaffold.of()` would
+/// not be able to find the [Scaffold], since the [Scaffold] is being created
+/// above the RaisedButton.
```
Comment (PixelToast) [inline packages/flutter/lib/src/widgets/basic.dart:7201]: We should explain that an error occurs if we call `Scaffold.of(context)` on our widget's own context, not that the snackbar simply wouldn't show.

I would write the following as something like "Since the `context` variable here is the context of our build method and not the RaisedButton itself, a call to `Scaffold.of(context)` will look for an ancestor in the wrong place."

File: packages/flutter/lib/src/widgets/basic.dart [hunk 2]
```diff
@@ -7160,13 +7160,88 @@ class KeyedSubtree extends StatelessWidget {
...
+///
+/// To resolve this problem, we use the [Builder] to return the [RaisedButton],
+/// and when we call `Scaffold.of(context)` in the RaisedButton's `onPressed`
+/// function, we are now using the context provided by the [Builder] and can
+/// successfully show the [SnackBar]:
```
Comment (PixelToast) [inline packages/flutter/lib/src/widgets/basic.dart:7206]: Could make this statement more general since it applies to more than just RaisedButton, i.e. "To access [ScaffoldState] properly in this example, we wrap its body in a [Builder] which gives us a context that is a child of the [Scaffold]."

File: packages/flutter/lib/src/widgets/basic.dart [hunk 3]
```diff
@@ -7160,13 +7160,88 @@ class KeyedSubtree extends StatelessWidget {
...
+/// {@tool sample --template=stateless_widget_material}
+/// Example of Scaffold.of With Builder
+/// ```dart
+/// Widget build(BuildContext context) {
+///   return Scaffold(
+///     appBar: AppBar(
+///       title: Text('Example'),
+///     ),
+///     body: Center(
+///       child: Builder(
```
Comment (PixelToast) [inline packages/flutter/lib/src/widgets/basic.dart:7217]: nit: This code could use some comments to draw our eyes closer to the important bits and hint to what is going on, for example the fact that we are intentionally shadowing the previous `context`.

**Comment by PixelToast** [comment]
Thanks for the PR!

**Review by pingbird — CHANGES_REQUESTED** [review]

File: packages/flutter/lib/src/widgets/basic.dart [hunk 4]
```diff
@@ -7160,13 +7160,88 @@ class KeyedSubtree extends StatelessWidget {
   Widget build(BuildContext context) => child;
 }
 
-/// A platonic widget that calls a closure to obtain its child widget.
+/// A simple widget that provides a specific [BuildContext] object to its child
+/// widget.
+///
+/// This [BuildContext] object is unique and is not the same object as
+/// the one provided by the framework in the `Widget build` method.
```
[the remark this answers is not in the document]
Comment (GroovinChip) [reply packages/flutter/lib/src/widgets/basic.dart:7167]: FWIW I was trying to write in a way that beginners could understand :)

File: packages/flutter/lib/src/widgets/basic.dart [hunk 5]
```diff
@@ -7160,13 +7160,88 @@ class KeyedSubtree extends StatelessWidget {
...
+///                content: Text('Have a snack!'),
+///              ),
+///            );
+///          },
+///        ),
+///      ),
+///    );
+///  }
+/// ```
+/// {@end-tool}
```
[the remark this answers is not in the document]
Comment (GroovinChip) [reply packages/flutter/lib/src/widgets/basic.dart:7195]: Fair enough

File: packages/flutter/lib/src/widgets/basic.dart [hunk 6]
```diff
@@ -7160,13 +7160,88 @@ class KeyedSubtree extends StatelessWidget {
...
 ///
 /// {@youtube 560 315 https://www.youtube.com/watch?v=xXNOkIuSYuA}
 ///
 /// See also:
 ///
 ///  * [StatefulBuilder], a platonic widget which also has state.
+///
+/// Legacy explanation:
+/// A platonic widget that calls a closure to obtain its child widget.
```
[the remark this answers is not in the document]
Comment (GroovinChip) [reply packages/flutter/lib/src/widgets/basic.dart:7244]: @RedBrogdon  requested I keep it :p this was the only way I could think to do it that preserved it yet kept it out of the way of the new explanation

File: packages/flutter/lib/src/widgets/basic.dart (hunk 3 above)
Comment (GroovinChip) [reply packages/flutter/lib/src/widgets/basic.dart:7217]: Noted

File: packages/flutter/lib/src/widgets/basic.dart (hunk 2 above)
Comment (GroovinChip) [reply packages/flutter/lib/src/widgets/basic.dart:7206]: Noted

File: packages/flutter/lib/src/widgets/basic.dart (hunk 1 above)
Comment (GroovinChip) [reply packages/flutter/lib/src/widgets/basic.dart:7201]: Noted

File: packages/flutter/lib/src/widgets/basic.dart (hunk 6 above)
[the remark this answers is not in the document]
Comment (PixelToast) [reply packages/flutter/lib/src/widgets/basic.dart:7244]: I saw the tweet, in that case we could incorporate `platonic widget` in the first paragraph somewhere

File: packages/flutter/lib/src/widgets/basic.dart (hunk 5 above)
[the remark this answers is not in the document]
Comment (Piinks) [reply packages/flutter/lib/src/widgets/basic.dart:7195]: Also, when using the template, this will be presented as a runnable app in the documentation, so we wouldn't want it to be a bad sample that crashes. :) 
Maybe you can use the `{@tool snippet` for this one, it won't be presented as an executable, and you can expand the discussion to point out why this won't work.

File: packages/flutter/lib/src/widgets/basic.dart (hunk 6 above)
[the remark this answers is not in the document]
Comment (Piinks) [reply packages/flutter/lib/src/widgets/basic.dart:7244]: Hmm, I've heard it both ways now. cc @timsneath and @sfshaza2 are we set on removing 'platonic'?

[the remark this answers is not in the document]
Comment (sfshaza2) [reply packages/flutter/lib/src/widgets/basic.dart:7244]: No, no, no. Andrew loves the English language and the use of "platonic" tickled his fancy. (He has a master's degree in poetry.) Ditch this historical info.

**Review by sfshaza2 — CHANGES_REQUESTED** [review]
@PixelToast, I like your feedback.  :D  I do have some changes I'd like to see.

**Comment by goderbauer** [comment]
@GroovinChip Do you have plans to address the latest feedback soon?

**Comment by GroovinChip** [comment]
> @GroovinChip Do you have plans to address the latest feedback soon?

Yes I do, my apologies - just got pulled in some other directions. Should be able to address tomorrow.

File: packages/flutter/lib/src/widgets/basic.dart (hunk 6 above)
[the remark this answers is not in the document]
Comment (GroovinChip) [reply packages/flutter/lib/src/widgets/basic.dart:7244]: Any word on the 'platonic' discussion?

File: packages/flutter/lib/src/widgets/basic.dart (hunk 4 above)
[the remark this answers is not in the document]
Comment (GroovinChip) [reply packages/flutter/lib/src/widgets/basic.dart:7167]: @PixelToast thoughts on marking this resolved?

[the remark this answers is not in the document]
Comment (PixelToast) [reply packages/flutter/lib/src/widgets/basic.dart:7167]: The feedback typically still stands until a new commit resolves it. If you are stuck then us and sfshaza could always collaborate in a google doc :)

**Comment by GroovinChip** [comment]
@goderbauer new commits, ready for review

File: packages/flutter/lib/src/widgets/basic.dart [hunk 7]
```diff
@@ -7153,7 +7153,83 @@ class KeyedSubtree extends StatelessWidget {
...
+///                content: Text('Have a snack!'),
+///              ),
+///            );
+///          },
+///        ),
+///      ),
+///    );
+///  }
+/// ```
+/// {@end-tool}
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/basic.dart:7191]: I think we should not post BAD EXAMPLES in our docs. Somebody might accidentally copy this.

File: packages/flutter/lib/src/widgets/basic.dart [hunk 8]
```diff
@@ -7153,7 +7153,83 @@ class KeyedSubtree extends StatelessWidget {
   Widget build(BuildContext context) => child;
 }
 
-/// A platonic widget that calls a closure to obtain its child widget.
+/// A platonic (simple) widget that provides a specific [BuildContext] object to its child
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/basic.dart:7156]: I thought we wanted to remove platonic? /cc @sfshaza2

File: packages/flutter/lib/src/widgets/basic.dart [hunk 9]
```diff
@@ -7153,7 +7153,83 @@ class KeyedSubtree extends StatelessWidget {
...
+/// widget.
+///
+/// This [BuildContext] object is unique and is not the same object as
+/// the one provided by the framework in the `Widget build` method.
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/basic.dart:7160]: "Widget build" method should not be in `. Also, Widget does not have a build method.

File: packages/flutter/lib/src/widgets/basic.dart [hunk 10]
```diff
@@ -19,53 +19,53 @@ import 'widget_span.dart';
 
 export 'package:flutter/animation.dart';
 export 'package:flutter/foundation.dart' show
-  ChangeNotifier,
-  FlutterErrorDetails,
-  Listenable,
-  TargetPlatform,
-  ValueNotifier;
+ChangeNotifier,
```
[the remark this answers is not in the document]
Comment (GroovinChip) [reply packages/flutter/lib/src/widgets/basic.dart:22]: I did revert the accidental formatting changes in [ef07f5d](https://github.com/flutter/flutter/pull/63584/commits/ef07f5d995a8cc039d5bebde2a4e00e5f7aa1c48)

File: packages/flutter/lib/src/widgets/basic.dart (hunk 7 above)
Comment (GroovinChip) [reply packages/flutter/lib/src/widgets/basic.dart:7191]: This was added due to previous feedback, if I could get a definitive resolution on this that would be great

File: packages/flutter/lib/src/widgets/basic.dart (hunk 8 above)
Comment (GroovinChip) [reply packages/flutter/lib/src/widgets/basic.dart:7156]: see previous comments

File: packages/flutter/lib/src/widgets/basic.dart (hunk 10 above)
[the remark this answers is not in the document]
Comment (goderbauer) [reply packages/flutter/lib/src/widgets/basic.dart:22]: This PR is still showing the formatting changes, though.

File: packages/flutter/lib/src/widgets/basic.dart (hunk 7 above)
Comment (goderbauer) [reply packages/flutter/lib/src/widgets/basic.dart:7191]: We should only show good practices in our docs and remove the bad example.

File: packages/flutter/lib/src/widgets/basic.dart (hunk 8 above)
Comment (goderbauer) [reply packages/flutter/lib/src/widgets/basic.dart:7156]: Which one?

File: packages/flutter/lib/src/widgets/basic.dart (hunk 10 above)
[the remark this answers is not in the document]
Comment (GroovinChip) [reply packages/flutter/lib/src/widgets/basic.dart:22]: Strange. I will check on it.

File: packages/flutter/lib/src/widgets/basic.dart (hunk 7 above)
Comment (GroovinChip) [reply packages/flutter/lib/src/widgets/basic.dart:7191]: Thank you

File: packages/flutter/lib/src/widgets/basic.dart (hunk 8 above)
Comment (GroovinChip) [reply packages/flutter/lib/src/widgets/basic.dart:7156]: https://github.com/flutter/flutter/pull/63584#discussion_r472406104

Comment (goderbauer) [reply packages/flutter/lib/src/widgets/basic.dart:7156]: I think we should remove "platonic" altogether, but I leave the last vote to @sfshaza2.

Comment (sfshaza2) [reply packages/flutter/lib/src/widgets/basic.dart:7156]: Yes, please remove platonic. It's too confusing, even for folks who know English pretty well.

**Comment by goderbauer** [comment]
(PR triage) @GroovinChip Do you still have plans to follow-up on this PR?

**Comment by GroovinChip** [comment]
I'm sorry. I do have plans but life is very complicated for me right now. Perhaps it is best if someone else takes over.

**Comment by goderbauer** [comment]
@HansMuller volunteered to take this one over.

**Comment by HansMuller** [comment]
I'm closing this PR now that #68151 has landed. Thanks for the contribution!
